src/contracts/verifier/contract/src/lib.rs

```diff
@@ -1,56 +1,394 @@
-        aurora: AccountId, //this can be safely set to "aurora"
-        contract_address: String,
+        aurora: AccountId, // This can be safely set to "aurora"
         nft_account_id: AccountId,
+        ic_remote_public_key: String,
+        contract_address: String,
     ) -> Self {
         Self {
             // This value only needs to be changed if you are running the aurora testnet locally
             aurora,
-            contract_address: aurora_sdk::parse_address(&contract_address).unwrap(),
             nft_account_id,
+            ic_remote_public_key: aurora_sdk::parse_address(&ic_remote_public_key)
+                .expect("ic_remote_public_key parse Error"),
+            contract_address: aurora_sdk::parse_address(&contract_address)
+                .expect("aurora_sdk parse Error"),
         }
     }
 
-    pub fn get_contract_state(&self) -> (AccountId, Address) {
-        return (self.nft_account_id.clone(), self.contract_address.clone());
+    /// Retrieves the current state of the contract
+    ///
+    /// # Returns
+    ///
+    /// * `(AccountId, Address, Address)` - Tuple containing the NFT account ID, contract address, and remote public key
+    pub fn get_contract_state(&self) -> (AccountId, Address, Address) {
+        return (
+            self.nft_account_id.clone(),
+            self.contract_address.clone(),
+            self.ic_remote_public_key,
+        );
+    }
+
+    /// Generates a Merkle tree from a vector of ProofResponse objects.
+    /// Each ProofResponse is hashed to create the leaves of the tree.
+    ///
+    /// # Arguments
+    ///
+    /// * `proof` - A string representing the proof
+    ///
+    /// # Returns
+    ///
+    /// * `String` - The Merkle root as a hexadecimal string
+    pub fn generate_merkle_tree(&self, proof: String) -> String {
+        // Convert each ProofResponse into a 32-byte hash to serve as a leaf in the Merkle tree.
+        let leaves: Vec<[u8; 32]> = vec![proof]
+            .iter()
+            .map(|proof_response| {
+                let proof_byte_content = proof_response.as_bytes();
+                Sha256::hash(proof_byte_content)
+            })
+            .collect();
+
+        // Construct the Merkle tree from the hashed leaves.
+        let tree: MerkleTree<Sha256> = MerkleTree::<Sha256>::from_leaves(&leaves);
+        let merkle_root = tree.root().expect("NOT ENOUGH LEAVES");
+        hex::encode(merkle_root)
     }
 
-    //make this payable and provide the metadata
+    /// Verifies an ECDSA signature against a given proof
+    ///
+    /// # Arguments
+    ///
+    /// * `proof` - A string representing the proof
+    /// * `signature` - A string representing the signature
+    ///
+    /// # Returns
+    ///
+    /// * `bool` - True if the verification is successful, false otherwise
+    pub fn ecdsa_verification(&self, proof: String, signature: String) -> bool {
+        let root_hash = hash_eth_message(self.generate_merkle_tree(proof));
```

Comment:
For maintainability, this should be done outside of the `ecdsa_verification`

The name of a function should describe exactly what's happening in the function.

Therefore generating the root hash (ecdsa signature message), should exist outside of the method to verify ecdsa (`ecdsa_verification`).

Just a note...